pkg/volume/cephfs/cephfs.go

```diff
@@ -0,0 +1,244 @@
+/*
+Copyright 2015 Google Inc. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cephfs
+
+import (
+	"os"
+	"strconv"
+	"strings"
+
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/api"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/types"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/util"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/util/exec"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/util/mount"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/volume"
+	"github.com/golang/glog"
+)
+
+// This is the primary entrypoint for volume plugins.
+func ProbeVolumePlugins() []volume.VolumePlugin {
+	return []volume.VolumePlugin{&cephfsPlugin{nil}}
+}
+
+type cephfsPlugin struct {
+	host volume.VolumeHost
+}
+
+var _ volume.VolumePlugin = &cephfsPlugin{}
+
+const (
+	cephfsPluginName = "kubernetes.io/cephfs"
+)
+
+func (plugin *cephfsPlugin) Init(host volume.VolumeHost) {
+	plugin.host = host
+}
+
+func (plugin *cephfsPlugin) Name() string {
+	return cephfsPluginName
+}
+
+func (plugin *cephfsPlugin) CanSupport(spec *api.Volume) bool {
+	if spec.VolumeSource.Cephfs != nil {
+		return true
+	}
+	return false
+}
+
+func (plugin *cephfsPlugin) GetAccessModes() []api.AccessModeType {
+	return []api.AccessModeType{
+		api.ReadWriteOnce,
+		api.ReadOnlyMany,
+		api.ReadWriteMany,
+	}
+}
+
+func (plugin *cephfsPlugin) NewBuilder(spec *api.Volume, podRef *api.ObjectReference) (volume.Builder, error) {
+	ep_name := spec.VolumeSource.Cephfs.EndpointsName
+	ns := api.NamespaceDefault
+	ep, err := plugin.host.GetKubeClient().Endpoints(ns).Get(ep_name)
+	if err != nil {
+		glog.Errorf("Cephfs: failed to get endpoints %s[%v]", ep_name, err)
+		return nil, err
+	}
+	glog.V(1).Infof("Cephfs: endpoints %v", ep)
+	return plugin.newBuilderInternal(spec, ep, podRef, mount.New(), exec.New())
+}
+
+func (plugin *cephfsPlugin) newBuilderInternal(spec *api.Volume, ep *api.Endpoints, podRef *api.ObjectReference, mounter mount.Interface, exe exec.Interface) (volume.Builder, error) {
+	id := spec.Cephfs.RadosUser
+	if id == "" {
+		id = "admin"
```

Comment:
Why is this defaulted here and not in the API defaults files?  There may be a legit reason, but I want to understand